src/Microsoft.Health.Checkpoints/Storage/StorageCheckpointClient.cs

```diff
@@ -33,78 +33,74 @@ public StorageCheckpointClient(BlobContainerClient containerClient, ILogger<Stor
                     {
-                        _logger.LogWarning($"Multiple blobs found for identifier {checkpointIdentifier}");
+                        checkpoint.LastProcessedIdentifier = lastProcessedIdentifier;
                     }
 
-                    foreach (BlobItem blobItem in blobPage.Values)
+                    if (string.IsNullOrWhiteSpace(lastProcessedDateTime) && string.IsNullOrWhiteSpace(lastProcessedIdentifier))
                     {
-                        DateTimeOffset lastEventTimestamp = DateTime.MinValue;
-
-                        if (blobItem.Metadata.TryGetValue(_lastProcessedDateTime, out var dt))
-                        {
-                            DateTimeOffset.TryParse(dt, null, DateTimeStyles.AssumeUniversal, out lastEventTimestamp);
-                            checkpoint.LastProcessedDateTime = lastEventTimestamp;
-                        }
-
-                        if (blobItem.Metadata.TryGetValue(_lastProcessedIdentifier, out var id))
-                        {
-                            checkpoint.LastProcessedIdentifier = id;
-                        }
-
-                        if (checkpoint.LastProcessedDateTime == DateTime.MinValue || checkpoint.LastProcessedIdentifier == null)
-                        {
-                            _logger.LogWarning($"No valid checkpoint found for {checkpointIdentifier}. Using default checkpoint of {checkpoint.LastProcessedDateTime}");
-                        }
-
-                        checkpoint.ETag = blobItem.Properties.ETag.ToString();
-                        checkpoint.Identifier = checkpointIdentifier;
+                        _logger.LogInformation($"No valid checkpoint found for {blobName}.");
+                        return null;
                     }
+
+                    checkpoint.ETag = blobItem.Properties.ETag.ToString();
+                    checkpoint.Identifier = checkpointIdentifier;
+                    checkpoint.Partition = partition;
                 }
             }
-            catch (Exception ex)
-            {
-                _logger.LogError(ex, ex.Message);
-                throw;
-            }
 
             return checkpoint;
         }
 
-        public virtual async Task<ICheckpoint> SetCheckpointAsync(ICheckpoint checkpoint)
+        public virtual async Task<ICheckpoint> SetCheckpointAsync(ICheckpoint checkpoint, CancellationToken token = default)
         {
             EnsureArg.IsNotNull(checkpoint);
             EnsureArg.IsNotNullOrWhiteSpace(checkpoint.Partition);
             EnsureArg.IsNotNullOrWhiteSpace(checkpoint.Identifier);
 
-            var lastProcessedDateTime = EnsureArg.IsNotNullOrWhiteSpace(checkpoint.LastProcessedDateTime.DateTime.ToString("MM/dd/yyyy hh:mm:ss.fff tt"), nameof(checkpoint.LastProcessedDateTime));
+            var lastProcessedDateTime = checkpoint.LastProcessedDateTime.ToString("MM/dd/yyyy hh:mm:ss.fff tt");
```

Comment:
Any particular reason to not use one of the standard ones? https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings

Reply:
Also, in this method, won't we be stripping away any offset that is set by the consumer and then possibly skewing the time since on the way out we assume universal? Seems like something like https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings#Roundtrip maybe more appropriate.

Reply:
That is a great point. Yes, roundtrip is a much better approach. The previous approach would have skewed the time. Updated to use roundtrip.